### PR TITLE
fix: Error properly on invalid and duplicate x:Name

### LINKED ISSUE #6198: [All Platforms] Duplicate ElementName (x:Name) should throw an error like UWP
## Current behavior

Duplicate ElementName (x:Name) doesn't throw an error like UWP

- For UWP:
_**XamlCompiler error WMC0047: Element name 'TwinElementName' is already used in this scope.**_

## Expected behavior

Duplicate ElementName (x:Name) should throw an error like UWP for all platforms.

## How to reproduce it (as minimally and precisely as possible)

```
	<StackPanel>
		<TextBlock x:Name="TwinElementName"
				   Text="Hello, world!" />
		<TextBlock x:Name="TwinElementName"
				   Text="Hello, world duplicate!" />
	</StackPanel>
```

## Environment

Nuget Package:

- [X] Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia
- [ ] Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia
- [ ] Uno.SourceGenerationTasks
- [ ] Uno.UI.RemoteControl / Uno.WinUI.RemoteControl
- [ ] Other: 

Nuget Package Version(s): Uno.UI 3.8.0-dev.449

Affected platform(s):

- [X] iOS
- [X] Android
- [X] WebAssembly
- [X] WebAssembly renderers for Xamarin.Forms
- [X] macOS
- [X] Skia
  - [X] WPF
  - [X] GTK (Linux)
  - [X] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

IDE:

- [ ] Visual Studio 2017 (version: )
- [X] Visual Studio 2019 (version: 16.10.0)
- [ ] Visual Studio for Mac (version: )
- [ ] Rider Windows (version: )
- [ ] Rider macOS (version: )
- [ ] Visual Studio Code (version: )

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #6198

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

No error on duplicate x:Name

## What is the new behavior?

Error on duplicate x:Name

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Review by Youssef1313 — COMMENTED** [review]
Can't find a good way to test this given https://github.com/unoplatform/uno/issues/6646

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 1]
```diff
@@ -3316,6 +3318,20 @@ void BuildCustomMarkupExtensionPropertyValue(IIndentedStringBuilder writer, Xaml
 			}
 		}
 
+		private void ValidateName(string? value)
+		{
+			// Nullable suppressions are due to https://github.com/dotnet/roslyn/issues/55507
+			if (!SyntaxFacts.IsValidIdentifier(value!))
+			{
+				throw new InvalidOperationException($"The value '{value}' is an invalid value for 'Name' property.");
+			}
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:3327]: UWP error message tells which character is invalid in the value. Let me know if you want me to make this update to the error message.

**Comment by davidjohnoliver** [comment]
From the CI it looks like the check is overly strict:

> Error UXAML0001: Processing failed for file C:\a\1\s\src\Uno.UI\Microsoft\UI\Xaml\Controls\SplitButton\SplitButton.xaml (System.InvalidOperationException: The name 'RootGrid' is already defined in this scope.

It's valid for the same `Name` to appear multiple times in the same XAML file in different 'scopes', eg within separate templates.

**Comment by Youssef1313** [comment]
> From the CI it looks like the check is overly strict:
> 
> > Error UXAML0001: Processing failed for file C:\a\1\s\src\Uno.UI\Microsoft\UI\Xaml\Controls\SplitButton\SplitButton.xaml (System.InvalidOperationException: The name 'RootGrid' is already defined in this scope.
> 
> It's valid for the same `Name` to appear multiple times in the same XAML file in different 'scopes', eg within separate templates.

@davidjohnoliver Is separate `ControlTemplate` the only exception where a duplication is allowed?

**Comment by davidjohnoliver** [comment]
> > From the CI it looks like the check is overly strict:
> > > Error UXAML0001: Processing failed for file C:\a\1\s\src\Uno.UI\Microsoft\UI\Xaml\Controls\SplitButton\SplitButton.xaml (System.InvalidOperationException: The name 'RootGrid' is already defined in this scope.
> > 
> > 
> > It's valid for the same `Name` to appear multiple times in the same XAML file in different 'scopes', eg within separate templates.
> 
> @davidjohnoliver Is separate `ControlTemplate` the only exception where a duplication is allowed?

`DataTemplates` as well I assume. That's all I can think of off the top of my head... @jeromelaban ?

Looking at the existing 'scope' logic in the code may give you an idea here, you may even be able to piggyback on it to simplify implementing your check.

**Comment by Youssef1313** [comment]
@davidjohnoliver I'm not 100% sure I'm doing it correctly. Can you check please? Thanks!

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 2]
```diff
@@ -151,6 +151,11 @@ internal partial class XamlFileGenerator
 		/// </summary>
 		private readonly bool _shouldAnnotateGeneratedXaml;
 
+		/// <summary>
+		/// Used to detect duplicate x:Name and report error for those.
+		/// </summary>
+		private HashSet<string> _names = new HashSet<string>();
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:157]: I think x:Name validations must be scoped to the proper "namescope". For instance, it's  valid to have two `DataTemplate` which contain the same `x:Name` member.

**Review by jeromelaban — CHANGES_REQUESTED** [review]

**Review by jeromelaban — COMMENTED** [review]
This looks better! Let's see what the CI says.

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs (hunk 1 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:3322]: That's not needed at this point. If the documentation tells which ones, let's redirect there.

**Comment by Youssef1313** [comment]
@jeromelaban This should be ready for review.

**Comment by Youssef1313** [comment]
Ping @jeromelaban for review

**Review by jeromelaban — APPROVED** [review]
Thanks for your contribution!

**Review by agneszitte-nventive — APPROVED** [review]

**Comment by agneszitte-nventive** [comment]
@jeromelaban / @Youssef1313 I think this PR can be merged no ?

**Pull request merged**